### ISSUE TITLE
Set limit for transcode cache

### ISSUE THREAD

**Issue by muhlinux** [jellyfin/jellyfin#3929]
I've moved my transcode directory to a ram disk which is nice, but capacity comes at a premium. It seems that the transcode cache just keeps growing, ideally it would be nice to have a setting for a simple size limit.

Thanks for considering, jellyfin is amazing!

**Comment by bingzhangdai** [comment]
comment just to active this issue. It would be great if we could have a smarter strategy for cache.

**Comment by fededim** [comment]
I am interested in this feature too, if you are using an SSD it will "eat" your drive lifespan quickly.

**Comment by joshuaboniface** [comment]
The cache *should* be automatically emptying itself of each transcoded video when playback ends. However unclean termination of the playback can interrupt this. There is a scheduled task (I forget the name off-hand) that will also clean this out regularly, but it is disabled by default since it, under normal circumstances, should not be needed.

Regarding sizing: Your transcoding directory needs enough space to transcode all file(s) being played at any given time. It thus does not make sense to "limit" it, because this will just cause you playback failures if you try to transcode a large file, or multiple streams at once. You need to give it sufficient storage space to transcode the largest possible video file you can give it, plus a bit of overhead just in case, and more if you will have multiple, simultaneous playbacks of different files.

Regarding SSD wear: This type of activity is *not* going to "eat" your SSD lifespan. SSD lifespans are exceedingly large compared to their sizes, even for cheap MLC/TLC/QLC NAND flash. This is no longer the late 2000's, and the amount of data we're talking is minimal compared to drive capacities and rated lifespans. If this is truly a concern, you should use a RAM-disk or invest in a hardier SSD. Limiting the size of the transcoding cache will do nothing here unless you want playback to forcibly stop when an arbitrary limit is reached.

**Comment by fededim** [comment]
Every SSD has its lifespan defined in TBW (Terabytes written), my drive has a decent value of 360 TBW. Suppose you are dealing with 4K HEVC video files of 15 GB size and suppose that Jellyfin transcodes 20 times a month, the yearly drive's wear in TBW becomes 20*15 GB*12 (months)=3.6 TB. It's 1% of my drive life yearly, but there are drives with a TBW of 100 and in this case it's the 3.6% yearly (in 5 years time it wears 18%). 

It's true that the transcode folder get cleaned as soon as you stop playing the video, but it keeps growing with all the transcoded streams since the beginning. In order to allow it to run in a ram disk (for those who do not want to wear out their drives) I think it would be useful to conceive an option which sets the maximum used disk space of the transcode folder and trims it also during play (if we are watching a movie of 2 hours it's high unlikely that we seek back to the beginning after 1 hour, if we do we can wait again for the re-transcoding on the seeked part). I do not how difficult this could be to implement but it should reduce to clean oldest files when reaching the maximum size and re-transcoding on seek (probably it's already the case when we seek forward in the movie).

**Comment by fededim** [comment]
In any case keep up the good work I found Jellyfin by chance, yet it it proving the best free, cross platform DLNA server with hardware transcoding on the market!

**Comment by bingzhangdai** [comment]
It makes sense to ensure “transcoding directory needs enough space to transcode all file(s) being played at any given time”, and disk lifetime is not a worry.

However, sometimes, we want to use ram disk to achieve the ultimate efficiency. I think we can configure two disks for transcoding cache, and Jellyfin can prefer to use first disk (e.g. ramdisk). If the first disk exceeds the limit, Jellyfin can fallback to the second one.

I believe it is a quite common optimization practice. The ram disk can be auto configured by Jellyfin and has a default (small) size. Of course, users can change it later. This issue can be classified enhancement, I think.

**Comment by fededim** [comment]
I have seen that there is in Dashboard / Advanced / Scheduled tasks menu there is "Clean transcode directory" which empties the files older than 1 day. Is it possibile to have it parametric in minutes ?

![image](https://user-images.githubusercontent.com/8364158/105495293-4a97c700-5cbc-11eb-9221-34af5cf3b23e.png)

**Comment by anthonylavado** [comment]
> Is it possibile to have it parametric in minutes ?

Possible? Yes. However, this task currently does not know whether or not a transcode is still in use. It just looks for files older than a certain timestamp (based on date created), and deletes them. When it was added, we figured a day (24 hours) was the best compromise, to help with users who may leave a session open (pausing and resuming as needed) throughout the day.

**Comment by fededim** [comment]
Ok not knowing whether or not a transcode is going on shouldn't be a problem. By deleting all files modified later than last x minutes (let's say 30) it would trim automatically the transcode directory during play (if no transcode is going it won't delete anything). Just a doubt: during a transcoding when you seek back it reuses the previously transcoded files or it re-transcodes from the seeked point ? If the latter is true, everything fits perfectly in the picture with such a simple change.

**Comment by anthonylavado** [comment]
@fededim If my understanding is correct, there are a few possible scenarios here. I'll focus on the two most likely.

1. Media playback starts, with transcoding/remuxing. You keep watching the stream without skipping forward. At some point, you decide to skip back. It will reuse the already transcoded file in the cache. (I believe this may also work if you skip forward within the boundaries of what's already been transcoded.)


2. Media playback starts, with transcoding/remuxing. At some point, you skip forward, but past the point that has already been transcoded by the server. The old session is stopped and a new one is spun up at this new timestamp. Progress resumes forward from this point with the same file. If you skip backwards past the original point, it will create a new transcode task.

**Comment by warchild-87** [comment]
I would love to have (full-functioning) RAMDISK-transcoding in Jellyfin!

**Comment by ich777** [comment]
I think I also post parts of my inital issue #5151 :

**System:**
 - OS: Unraid 6.9.0RC2
 - Virtualization: Docker
 - Clients: Browser, Android
 - Browser: Firefox
 - Jellyfin Version: latest from DockerHub
 - Playback: Transcode
 - Installed Plugins:none (only the standard ones)
 - Reverse Proxy: none (local access)
 - Base URL: SERVERIP:8096
 - Networking: Bridge
 - Storage: local

**Steps to Reproduce**
1. Add `--mount type=tmpfs,destination=/tmp,tmpfs-size=500000000` to your `docker run` command (this creates a tmpfs in the /tmp directory, in this example with 500MB).
2. Go to your settings and set the directory for transcoding to '/tmp' and click save.
3. Open up a video file and force it to transcode (should be a big file so that the transcoding buffer is filled up quickly)
4. After the directory is filled up it ends transocding and won't transcode any further
5. The video plays fine until it reaches the timecode where it ended transocding because it's out of space and the video is stale
    (You can still click on the 'Back' button to go back to the Menu, after doing that the transcoding directory is cleaned)

**Expected behavior**
Playback shouldn't stop after the directory is filled up instead it should slowly delete the oldest file in the directory that was transcoded so that it playback never stops.

**Additional context**
I recently came across [this thread](https://forums.unraid.net/topic/35878-plex-guide-to-moving-transcoding-to-ram/) in the Unraid forums where it was discussed how to set up a tmpfs with a mount from the host '/tmp/ to the container '/tmp' directory, but it was also discussed that this is not an ideal solution because when the RAM is filled up the server crashes, so a script was written to limit the amount of RAM that this directory inside the container can use and not cause a server crash ([this was the link to the script](https://forums.unraid.net/topic/35878-plex-guide-to-moving-transcoding-to-ram/?do=findComment&comment=894460) but it is now changed to the easier method like described above to put the command in the `docker run` command) to not stress the hard drives when transcoding.
The problem is that this works flawlessly with Plex (Plex deletes the oldest transcoded file if it reaches a certain level of use in the directory, about 90%, and a certain timecode) but Jellyfin will stale if it runs out of space and the playback stops after it reaches the last file that was transcoded in the transocding directory.


Would also love to see this feature coming to Jellyin.

**Comment by lunainvictum** [comment]
The statement about the SSD's wear is only semi-correct. This is because hardly any SSDs "break" or become unusable because they are worn out, but break for other reasons like material defects, short circuits or other unknown reasons. There is a study by Facebook and Google, which once examined this, and why they continue to take normal hard drives or RAM for caching. As far as I can remember, the main reason for the defects was not that the memory chips were defective or worn out, but because the controller broke - and that only far, very far, above the specified TBW of the manufacturer. 
With this study, they wanted to evaluate whether it is possible to reliably predict when an SSD dies, which didn't work as far as I can remember. The study is approximately from the year 2010. 

Even if you assume that a user (like I'm one, for example, and now posting here) must transcode over 100GB a day, an SSD with 500GB (which costs only 40-50€ here in Germany, for example) with 150TBW (and that's the 40€ variant) lasts over 4 years. And honestly, my server doesn't even last that long, because it tends to be replaced because new, faster interfaces are available. But until that happens, the SSD has already been sold 2 years earlier, because it was exchanged for a larger one, because of a constantly growing library. 

Well, never mind. Everyone has a different opinion on the subject.

Which is why I'm here in any case, is the fact that I'm unfortunately very lazy and I do not care if the media "in the right format" and take everything I can get. That's why my server has a Nvidia 1070ti in it with which is transcoded. And I run 24/7 some TV series randomly on 3 Raspberry Pi's. 

And it doesn't matter if you shutdown "properly" by pressing stop or if you just take the Raspberry Pi's off power and do a "hard shutdown". There is always the transcoding data (or remains of it) left, which then adds up.

I now have to empty the transcode folder every 4 hours and restart jellyfin + all clients (transcode folder has a capacity of 50GB). 

The argument that the transcode data remains because it was not "cleanly terminated" cannot be correct. For the reason (I can make a video) that if you watch the Jellyfin admin panel you can see to the millisecond exactly which episode which Raspberry Pi is playing or if a Raspberry Pi has just been killed (and therefore logically no more playback). This means that the server knows exactly what is happening, and should then also react and delete the respective data. 

In this thread, however, it was claimed that the data always remains when a session has not been properly terminated and thus you should see in the admin panel "phantom playback time".

This problem should not be seen as an "enhancement" but as a critical bug (for those who have to transcode a lot).

For now, I have solved the problem with me so that when starting the jellyfin server a cronjob is done, which deletes all files. Since my server restarts every day 4 o'clock in the morning anyway. 

I hope this long text is understandable, this was translated from German into English with deepl.

Translated with www.DeepL.com/Translator (free version)

**Comment by TheHellSite** [comment]
Any news on this?

I too think that this is more a bug than a feature request.
Even with a transcoding moved to RAM I would need a huge amount of RAM just to be safe.

I just did a quick test by "downscaling-transcoding" an 11 Mbps movie to 10 Mbps.
The movie file size is ~20 GB.
It uses ~100MB of disk size in about 20 seconds.

**Comment by peterge1998** [comment]
This is the 2nd time now my server becomes unable to login because /var/lib/jellyfin/transcode gets full of .ts files, so that my 30 GB disk runs 100% full...
How did you try to solve this problem?
A cronjob?

**Comment by TheHellSite** [comment]
> This is the 2nd time now my server becomes unable to login because /var/lib/jellyfin/transcode gets full of .ts files, so that my 30 GB disk runs 100% full... How did you try to solve this problem? A cronjob?

I am using my HDD based NAS (samba share) for this...
It is more than fast enough and has terabytes of storage. I simply mount it to "/var/lib/jellyfin/transcodes".

However for me this is also just a workaround.

**Comment by zynexiz** [comment]
Just had the same issue couple of times now, last time 5 mins ago. Had to manually empty transcoding directory. Think I have to create some kind of cron jobb that empty it every night, the scheduled task thing in Jellyfin itself doesn't do anything (have tried setting it to run ever 3 hours, nothing)

**Comment by peterge1998** [comment]
> Just had the same issue couple of times now, last time 5 mins ago. Had to manually empty transcoding directory. Think I have to create some kind of cron jobb that empty it every night, the scheduled task thing in Jellyfin itself doesn't do anything (have tried setting it to run ever 3 hours, nothing)

Where is the integrated task you mean?
Did you set up the job yet? Did it work? Can you post it please?

**Comment by zynexiz** [comment]
> > Just had the same issue couple of times now, last time 5 mins ago. Had to manually empty transcoding directory. Think I have to create some kind of cron jobb that empty it every night, the scheduled task thing in Jellyfin itself doesn't do anything (have tried setting it to run ever 3 hours, nothing)
> 
> Where is the integrated task you mean? Did you set up the job yet? Did it work? Can you post it please?

Jellyfin itself have scheduled tasks, where one is to empty the transcodoing directory, but it doesn't do that it seems. I haven't set up my own cron jobb yet, lot of stuff to do in my university atm..

**Comment by opeeba** [comment]
The Clean Transcodes Directory task only deletes files more than one day
old according to its description. It wouldn't matter how frequently it is
run if the existing files are less than a day old.

**Comment by HNJAMeindersma** [comment]
> This is the 2nd time now my server becomes unable to login because /var/lib/jellyfin/transcode gets full of .ts files, so that my 30 GB disk runs 100% full... How did you try to solve this problem? A cronjob?

Exact same issue here. I'll try to mount a SMB share from my NAS as a larger cache on HDD instead of SSD. I would greatly appreciate a set maximum amount of transcode cache, including more active housekeeping of it. For now it almost feels like a deal breaking issue to use Jellyfin, since it just stops in the middle of a movie.

**Comment by 0xF5F5F5** [comment]
I was just enjoying Jellyfin, after setting the transcode folder to a ramdisk created by ImDisk, but playback stopped. I looked to see if the ramdisk was full, and it was. I thought Jellyfin would just delete the oldest transcoded parts, but instead it appears to die.

Is there a solution?

**Comment by 0xF5F5F5** [comment]
>For now it almost feels like a deal breaking issue to use Jellyfin, since it just stops in the middle of a movie.

I agree. I feel like writing a script that checks if my ramdisk is getting full, and if it is, it deletes a fixed number of the oldest files in it. Once it works, I'll use Windows task scheduler to have it run every few minutes. Load should be slow because, at the very least, the script only checks the free space on the ramdisk every few minutes

**Comment by zynexiz** [comment]
As a workaround, I created a bash script that checks the available space on the drive and purge the transcode directory if the space left in lower than the threshold and added a cronjobb that runs it ever 20 mins.

/usr/bin/jellyfin-purge (chmod +x to make it executable)
```
usedspace=$(df / --output='pcent' | grep -o "[0-9]*")

if [ $usedspace -gt 78 ]; then 
        rm -r /var/lib/jellyfin/transcodes/* 2> /dev/null
        printf "Purged Jellyfin transcode directory, space left on drive $usedspace%%" | systemd-cat --identifier=jellyfin
fi
```
Change 78 to some percentage that will fit your needs. Then call the script thru a cronjobb.
Run sudo crontab -e -u jellyfin and add the following lines;
`0,20,40 * * * * /usr/bin/jellyfin-purge # Purge transcode directory in jellyfin`
This will run the script every 20 mins (change this if you want to run it more or less often).

**Comment by joshuaboniface** [comment]
There already is a scheduled task for this (clean up transcoding), which IIRC is on nightly by default.

That said, transcoding temp should clean itself up automatically when playback stops.

If the issue is it running out of space *during* playback, then that's its own optimization that would need to be investigated.

**Comment by NephtonLabs** [comment]
Left IPTV running and the server filled up and died. A hard cap on the size this folder can grow to would avoid this problem.

**Comment by nichobi** [comment]
> If the issue is it running out of space _during_ playback, then that's its own optimization that would need to be investigated.

I think this is the more fundamental issue in the long term. For people using a RAM disk as transcoding storage, storing the entire film at once will always run the risk of exceeding memory.  

My own experience with 16 GB of RAM has been that it works flawlessly most of the time, but if someone tries to watch a 30 GB film and requires transcoding, it will give up halfway through and potentially slow the whole server to a crawl if the memory is entirely full.

Keeping the entirety of a 2.5 hour film in memory just to watch it through to the end doesn't seem like optimal behaviour to me. A short backwards buffer (10 minutes or so?) ought to be plenty to avoid retranscoding in most cases of skipping backwards.

**Comment by deddc23efb** [comment]
I've addressed this with a cronjob that deletes files in the transcodes folder that are older than 2 hours.

**Comment by ich777** [comment]
> I've addressed this with a cronjob that deletes files in the transcodes folder that are older than 2 hours.

But that doesn't solves the issue at all, if someone has multiple users and want to set a RAM disk with 1GB or even less that is simply not enough and also won't fix this.

**Comment by ocervell** [comment]
Thinking of the same migration, would be nice to modify the existing embedded task to @zynexiz script maybe.

**Comment by osscombat** [comment]
In case of LiveTV streaming this issue is a show stopper - almost all free space is simply filled up as long as you watch the stream. This doubles up with every other user that decided to watch LiveTV with transcoding. The files are not being overwritten, the transcode just stops.

**Comment by dkuchay** [comment]
I'm suddenly here because I installed Jellyfin about 4 hours ago and my server went belly up. After some investigating its the transcodes file at over 80G! For the love of god please put in bold caps in the description of this program "will eat SSD's or consume your whole SSD overnight" until this issue gets fixed. @[joshuaboniface] not one user on this planet will ever guess that any program out there will consume 80+g's in a few hours. This was not well described in the apps desciption. This is a significant bug. I was first thinking that I wont be able to use jellyfin until I can afford a much much larger SSD. THANK YOU to Zunexiz!!! I dont remember uncleanly exiting anything. What could I watch in one night (less than 4 hours since installing) that could equal 80GB???!! Bug

**Comment by cvium** [comment]
Plex has proper throttling which allows them to use the built-in segment limiter in ffmpeg (whether they use that or delete manually idk) :)

**Comment by dkuchay** [comment]
More information: TV is Sony XBR with Android TV. XBR49X800EJellyfin for Android TV installed. Server is Xpenology OS which is unofficial synology. Synology Community Jellyfin installed v10.7.7.I set the hdhomerun to transcode in an effort to offload from NAS. NAS has an old GTX750 Nvidia. Replacement 1050TI on route. Plex did not struggle on same server like Jellyfin seems to. Upgrade card was ordered after watching some looping on streams in jellyfin. The transcode file came from watching a few hours of streaming from HDHomerun with transcode set on HDHomerun.My plex directory on the same server is at ***@***.***:/volume5$ sudo du -h --max-depth=1 7.0G    .***@***.***     .***@***.***       .***@***.***    ./Plex0       .***@***.***    .What I find interesting is that transcoding is even taking place at all. I believe that I am sending mpeg-2 with two channel audio to a modern TV.... Are modern smart tv's not capable of playing files from an HDHomerun? I would think that transcode would come into play when changing wrapper or dimensions on the video. Not when sending files formatted for a TV to a TV?  I will remove trancode from the tuner setup and test again. Just to make sure mpeg-2 is sending. I am guessing Plex can discover the native capabilities of the connected TV more so than Jellyfin? Another topic but the SSL file I used in Plex for connecting IOS over SSL is not allowing Jellyfin to connect on same method. On Mar 31, 2022, at 6:05 AM, Claus Vium ***@***.***> wrote:Plex has proper throttling which allows them to use the built-in segment limiter in ffmpeg (whether they use that or delete manually idk) :)—Reply to this email directly, view it on GitHub, or unsubscribe.You are receiving this because you commented.Message ID: ***@***.***>

**Comment by kaysond** [comment]
Also ran into this issue. Some kind of improved optimization or a cache limit would be great!

**Comment by Write** [comment]
This seems to be a huge issue

**Comment by alpe12** [comment]
Still nothing? Big issue.

**Comment by Henkow** [comment]
I also find this to be a big problem. Having the transcoding stay a couple minutes ahead of the movie would be great. If people want to transcode the whole movie to ramdisk to allow for seamless transition then let them, but also allow for a cache limit.

**Comment by DomiStyle** [comment]
I added support for segment deletion in the branch here:

https://github.com/DomiStyle/jellyfin/tree/segment-deletion
https://github.com/DomiStyle/jellyfin-web/tree/segment-deletion

Would be cool if some people could test with different clients to see if there are any issues.
Works fine for me on Linux with Live TV and regular files, VAAPI transcoding and software transcoding.

I also made the seconds to buffer configurable, so feel free to play around with it. The defaults seem to make sense in my testing though. By default Jellyfin will throttle after 3 minutes and will keep segments for 3.3 minutes.

In order to enable it you will have to build Jellyfin and Jellyfin Web from the branch above and enable it in the encoder settings:
![Screenshot 2022-11-14 at 10-30-57 Jellyfin](https://user-images.githubusercontent.com/975131/201628202-bd5dd047-6ce4-4a1c-aaa3-b93e5c1f217d.png)

**Comment by MaxXor** [comment]
My `/var/lib/jellyfin/transcode` path is empty, does that mean my jellyfin server is not transcoding anything?

**Comment by nyanmisaka** [comment]
> I added support for segment deletion in the branch here:
> 
> https://github.com/DomiStyle/jellyfin/tree/segment-deletion https://github.com/DomiStyle/jellyfin-web/tree/segment-deletion
> 
> Would be cool if some people could test with different clients to see if there are any issues. Works fine for me on Linux with Live TV and regular files, VAAPI transcoding and software transcoding.
> 
> I also made the seconds to buffer configurable, so feel free to play around with it. The defaults seem to make sense in my testing though. By default Jellyfin will throttle after 3 minutes and will keep segments for 3.3 minutes.
> 
> In order to enable it you will have to build Jellyfin and Jellyfin Web from the branch above and enable it in the encoder settings: ![Screenshot 2022-11-14 at 10-30-57 Jellyfin](https://user-images.githubusercontent.com/975131/201628202-bd5dd047-6ce4-4a1c-aaa3-b93e5c1f217d.png)

This should work with HLS only, progressive playback is not affected.
Also, it would be great if you can open a PR for it.

**Comment by DomiStyle** [comment]
> This should work with HLS only, progressive playback is not affected.

Correct, couldn't quite figure out when Jellyfin uses progressive playback. All the media that requires this feature used HLS for me.

I assume it's for players that don't support HLS?

> Also, it would be great if you can open a PR for it.

Will do so later today.

**Comment by MaxXor** [comment]
> > My /var/lib/jellyfin/transcode path is empty, does that mean my jellyfin server is not transcoding anything?
> 
> Is that the same path as listed in Playback -> Transcode path?
> 
> You can force the bitrate to something low to make sure.

It is indeed the same path. Does that maybe mean that my end devices dont need any transcoding?

**Comment by DomiStyle** [comment]
> Does that maybe mean that my end devices dont need any transcoding?

Most likely, try reducing the bitrate in the player to something low and see if it transcodes then. You can also check the state in the Administration dashboard per stream.

**Comment by DomiStyle** [comment]
>  The above solution for deleting old segments wouldnt be sufficient because lets say 4 users are using it at once, there will be no old segments to delete as they would all be active. 

I don't really get what you mean by this, the segments will still be deleted since it opens separate transcoding sessions for each playback.

Once the PR is merged you can just enable segment deletion in the settings.

**Comment by RTUnit** [comment]
While we are waiting for this feature to be released, I have created a workaround for cleaning up the RAM drive. Thought to share it here if someone wants to re-use it. Below script will check the used space in transcoding directory (I use **/config/transcodes**), if used space is **>= 85%** then it will delete oldest 20 files every 1 second, if **>= 75%** then it will delete oldest 10 files every 30 seconds. The script will also delete all 0 size files from transcoding directory.

You can set it up using cron, but since I run Jellyfin server on docker then I needed to create wrapper for ffmpeg and launch the cleanup script in background using command: `sh /config/ffmpeg/transcode.cleanup.sh &` just before launching ffmpeg. The script will create PID file (/config/ffmpeg/transcode.cleanup.pid), so the next time another instance is launched, it will kill the old one.

**/config/ffmpeg/transcode.cleanup.sh**
```
#!/bin/sh

# check if PID file exists (clean up process may be already running)
if [ -f /config/ffmpeg/transcode.cleanup.pid ]; then

   OLD_PID=$(head -1 /config/ffmpeg/transcode.cleanup.pid)
   echo "Previous PID file found with PID=$OLD_PID"

   # check if PID belongs to clean up process (or it might be terminated and another process working with the same PID)
   # command line should match to: /bin/sh/config/ffmpeg/transcode.cleanup.sh

   OLD_CMD=$(cat /proc/$OLD_PID/cmdline 2>&1 | tr -d '\0')
echo $OLD_CMD
   if [ "$OLD_CMD" = "/bin/sh/config/ffmpeg/transcode.cleanup.sh" ] || [ "$OLD_CMD" = "sh/config/ffmpeg/transcode.cleanup.sh" ]; then

     echo "PID $OLD_PID belongs to clean up process, killing it"
     # kill the clean up process using PID stored in /bin/sh/config/ffmpeg/transcode.cleanup.pid file
     kill $OLD_PID
   else
     echo "PID $OLD_PID is not clean up process, will not touch this process"
   fi;
fi

echo $$ > /config/ffmpeg/transcode.cleanup.pid

# check used space for /config/transcodes,
# if >= 85% used then delete 20 oldest *.ts files, sleep for 1 seconds
# if >= 75% used then delete 10 oldest *.ts files, sleep for 30 seconds
while true; do
   SPACE_USED=$(df /config/transcodes -H --output=pcent | sed '1d;s/.$//')
   if [ $SPACE_USED -ge 85 ]; then
      echo Space used is $(df /config/transcodes -H --output=pcent | sed '1d;s/.$//')'%', deleting 20 oldest files...
      ls -1t /config/transcodes/*.ts | tail -n 20 | xargs rm -f
      SLEEP_SEC=1
   else
      if [ $SPACE_USED -ge 75 ]; then
         echo Space used is $(df /config/transcodes -H --output=pcent | sed '1d;s/.$//')'%', deleting 10 oldest files...
         ls -1t /config/transcodes/*.ts | tail -n 10 | xargs rm -f
      else
         echo Space used is $(df /config/transcodes -H --output=pcent | sed '1d;s/.$//')'%'
      fi;
      SLEEP_SEC=30
   fi;
   # delete all 0 size files in transcodes directory (created with 0 size due to lack of free space)
   # such files will cause HLS player to stall during playback
   find /config/transcodes -size 0 -type f -delete
   sleep $SLEEP_SEC
done
```

**/config/ffmpeg/ffmpeg.wrap**
```
#!/bin/bash -
#description     :  This script is a wrapper for jellyfin ffmpeg

#ffmpeg="/usr/bin/ffmpeg"
ffmpeg="/usr/lib/jellyfin-ffmpeg/ffmpeg"

#Add quotes for arguments with "spaces" and "parentheses")
for x in "${@}" ; do
    # try to figure out if quoting was required for the $x (Arguments)
    if [[ "$x" != "${x%[[:space:]]*}" ]] \
    || [[ "$x" != "${x%(*}" ]] ; then
        x="\""$x"\""
    fi
    _args=$_args" "$x
done

# Execute transcode cleanup process, which will remove oldest *.ts files when space is used for more than 75%
# If parallel ffmpeg process will be launched then previous clean up process will be killed
sh /config/ffmpeg/transcode.cleanup.sh &

# Execute wrapper
eval $ffmpeg $_args

exit $?
```

Issues: if you run the script manually then your instance won't be killed by the next instance that is started by ffmpeg wrapper. Don't know why.

**Comment by dute533** [comment]
Inspired by @zynexiz I'm using the following bash script: 

`#!/bin/bash

set -o pipefail
decodedir=/dev/shm/transcodes

while true; do
  while inotifywait -e modify $decodedir; do
    usedspace=$(df $decodedir --output='pcent' | grep -o "[0-9]*")
    if [ $usedspace -gt 80 ]; then
      rm $decodedir/"$(ls $decodedir -t | tail -1)"
    fi
  done
done
`

Seems to work well so far, running it from a systemd service as the jellyfin user. 
I think Throttle Transcodes is probably necessary, to avoid removing files you're still watching.

**Comment by Bond-009** [comment]
10.9 will have an option to auto delete segments
#8744

**Comment by tenshiism** [comment]
> -e -u jellyfin and add the following lines (or -u root depending on you installation);

hi i'm not sure what this means as I just started using linux and if I started explaining the parts you said I don't understand it would take forever. Has this comment been implemented in jellyfin yet or is there a command I can run to download and install it?

**Comment by MaxXor** [comment]
Don't worry about the cronjob to manually purge the jellyfin transcode directory, Jellyfin 10.9 will be released soon and fixes this.

**Comment by lzap** [comment]
Just for the record, if you are on Linux, have a plenty of RAM like me and want to prevent the theoretical SSD wear, it is super easy:

    sudo $EDITOR /etc/fstab

Add this line (change `4g` for maximum limit, 4 GiB in this case):

```
tmpfs /var/lib/jellyfin/transcodes tmpfs defaults,size=4g 0 0
```

    sudo systemctl daemon-reload

And

    sudo mount -a

That is all.

Edit: This comment is somewhat irrelevant to the issue, I would love to be able to configure the cache size as well. I just thought I would share this, it is a way to limit the size in a forceful way.

**Comment by bakhtiyarneyman** [comment]
> Just for the record, if you are on Linux, have a plenty of RAM like me and want to prevent the theoretical SSD wear, it is super easy:

This works great until it doesn't. If you hit the ceiling because you are seeking too much, the player will stall.

**Comment by doublep21** [comment]
yes this issue needs to be reopened , i need it too

**Comment by illsk1lls** [comment]
I currently am not watching anything and my transcode directory (ram) is holding ~72gb of ram, and its supposed to be limited to 64gb via the settings, its showing as orange in the dashboard, the scheduled task to clean the "folder" every 30 minutes as i have it set isnt working. Its also strange the limit is being ignored.

I am using TrueNAS Scale, with the Jellyfin app

**Comment by akrepgeneral** [comment]
I don't understand why you're ignoring this problem. As you can see, it's important for many of us and it's still unresolved after 5 years.

**Comment by RTUnit** [comment]
As with any other communittee software, anyone can contribute to develop this new feature. It looks as there is no developer who is interested to implement it, most likely because developers can find ways to workaround this problem by scheduling clean-up scripts.

Others could re-use the scripts to achieve the same. For example, if you are on Linux, you can use my script for cleanup: https://github.com/RTUnit/Jellyfin-Transcodes-cleanup. If you are on Windows then maybe someone else will help to provide something similar to re-use.
